crates/eframe/src/web/text_agent.rs

```diff
@@ -1,237 +1,177 @@
+        runner_ref.add_event_listener(&input, "compositionupdate", on_composition_update)?;
+        runner_ref.add_event_listener(&input, "compositionend", on_composition_end)?;
 
-    body.append_child(&input)?;
+        Ok(Self {
+            input,
+            prev_ime_output: Default::default(),
+        })
+    }
 
-    Ok(())
-}
+    pub fn move_to(
+        &self,
+        ime: Option<egui::output::IMEOutput>,
+        canvas: &web_sys::HtmlCanvasElement,
+    ) -> Result<(), JsValue> {
+        // don't move the text agent on mobile
+        if is_mobile() {
+            return Ok(());
+        }
 
-/// Focus or blur text agent to toggle mobile keyboard.
-pub fn update_text_agent(runner: &AppRunner) -> Option<()> {
-    use web_sys::HtmlInputElement;
-    let window = web_sys::window()?;
-    let document = window.document()?;
-    let input: HtmlInputElement = document.get_element_by_id(AGENT_ID)?.dyn_into().unwrap();
-    let canvas_style = runner.canvas().style();
+        // don't move the text agent unless the position actually changed:
+        if self.prev_ime_output.get() == ime {
+            return Ok(());
+        }
+        self.prev_ime_output.set(ime);
 
-    if runner.mutable_text_under_cursor {
-        let is_already_editing = input.hidden();
-        if is_already_editing {
-            input.set_hidden(false);
-            input.focus().ok()?;
+        let Some(ime) = ime else { return Ok(()) };
 
-            // Move up canvas so that text edit is shown at ~30% of screen height.
-            // Only on touch screens, when keyboard popups.
-            if let Some(latest_touch_pos) = runner.input.latest_touch_pos {
-                let window_height = window.inner_height().ok()?.as_f64()? as f32;
-                let current_rel = latest_touch_pos.y / window_height;
+        let ime_pos = ime.cursor_rect.left_top();
+        let canvas_rect = canvas.get_bounding_client_rect();
+        let new_pos = ime_pos + egui::vec2(canvas_rect.left() as f32, canvas_rect.top() as f32);
 
-                // estimated amount of screen covered by keyboard
-                let keyboard_fraction = 0.5;
+        let style = self.input.style();
+        style.set_property("top", &format!("{}px", new_pos.y))?;
+        style.set_property("left", &format!("{}px", new_pos.x))?;
 
-                if current_rel > keyboard_fraction {
-                    // below the keyboard
+        Ok(())
+    }
 
-                    let target_rel = 0.3;
+    pub fn set_focus(&self, on: bool) {
+        if on {
+            self.focus();
+        } else {
+            self.blur();
+        }
+    }
 
-                    // Note: `delta` is negative, since we are moving the canvas UP
-                    let delta = target_rel - current_rel;
+    pub fn has_focus(&self) -> bool {
+        let active_element = super::focused_element();
+        self.input.clone().dyn_into().ok() == active_element
```

Comment:
This will return `true` if the left side errors (i.e. `.ok()` results in `None`)